Add RangeSet and discontiguous collection operations

natecook1000 opened the pull request:
This adds the `RangeSet` and `DiscontiguousSlice` types, as well as collection operations for working with discontiguous ranges of elements. This also adds a `COWLoggingArray` type to the test suite to verify that mutable collection algorithms don't perform unexpected copy-on-write operations when mutating slices mid-operation.

Still to do:
- [x] Add documentation for `RangeSet` set algebra methods (since we aren't conforming, we don't inherit the docs from the protocol)
- [x] Add versioning attributes

natecook1000 committed:
Add RangeSet and discontiguous collection operations

This adds the RangeSet and DiscontiguousSlice types, as well as collection
operations for working with discontiguous ranges of elements. This also adds
a COWLoggingArray type to the test suite to verify that mutable collection
algorithms don't perform unexpected copy-on-write operations when mutating
slices mid-operation.

natecook1000 committed:
Fix broken invariant in single-range initializer

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 committed:
Merge branch 'master' into nc_rangeset

natecook1000 committed:
Revise proposed RangeSet & collections API

natecook1000 committed:
Revise proposed RangeSet & collections API for proposal v3

futurejones commented:
Can one of the admins verify this patch?

natecook1000 committed:
Merge branch 'master' into nc_rangeset

natecook1000 committed:
[stdlib] Adopt the custom RangeSet storage type

natecook1000 committed:
Update copyright notices

natecook1000 committed:
Switch to standard library preconditions

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 committed:
Merge branch 'master' into nc_rangeset

natecook1000 committed:
Remove ExpressibleByArrayLiteral conformance

natecook1000 committed:
Add availability attributes to RangeSet &co

natecook1000 committed:
Simplify RangeSetStorage's single-range case

Just use a `Range<T>` as is for the single-range case. If there's
a benefit to making range sets more compact we can move toward that
instead.

natecook1000 committed:
Fill in remaining RangeSet documentation

natecook1000 commented:
@swift-ci Please smoke test

xwu commented on stdlib/public/core/RangeSet.swift:
```diff
@@ -10,7 +10,23 @@
-/// A set of ranges of any comparable value.
+/// A set of any comparable value, represented by ranges.
+///
+/// You can use a range set to efficiently represent a set `Comparable` values
```
“set __of__”

xwu commented on stdlib/public/core/RangeSet.swift:
```diff
@@ -0,0 +1,594 @@
+//
+//===----------------------------------------------------------------------===//
+
+/// A set of any comparable value, represented by ranges.
```
“any comparable value__s__”?

natecook1000 committed:
Minor revisions to RangeSet type docs

natecook1000 committed:
Update test fixtures for RangeSet subscript

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 committed:
Fix a documentation type

natecook1000 committed:
Remove RangeSet.contains(_: Range) method

This method was removed from the final approved SE proposal.

natecook1000 committed:
Switch remaining preconditions to internal stdlib versions

natecook1000 committed:
Inversion of empty rangeset should be the whole bounds

natecook1000 committed:
Inlinability annotations

milseman commented on stdlib/public/core/RangeSet.swift:
```diff
@@ -0,0 +1,581 @@
+  ///
+  /// - Complexity: O(*n*), where *n* is the number of ranges in the range
+  ///   set.
+  public mutating func insert<C>(_ index: Bound, within collection: C)
```
This is a candidate for effective `inlinable`, depending on performance needs. Doing so will drop generality from Collection to just our Bound. Likely big benefit, write a benchmark to demonstrate.

milseman commented on stdlib/public/core/RangeSet.swift:
```diff
@@ -0,0 +1,581 @@
+  ///   - index: The index to include in the range set. `index` must be a
+  ///     valid index of `collection` that isn't the collection's `endIndex`.
+  ///   - collection: The collection that contains `index`.
+  public init<S, C>(_ indices: S, within collection: C)
```
This could be a candidate for `inlinable`, depending on need. It's a bigger code size tradeoff than `insert` below.

milseman commented on stdlib/public/core/RangeSet.swift:
```diff
@@ -0,0 +1,581 @@
+  ///
+  /// - Complexity: O(*n*), where *n* is the number of ranges in the range
+  ///   set.
+  public mutating func remove<C>(_ index: Bound, within collection: C)
```
`inlinable` could be a win.

milseman commented on stdlib/public/core/RangeSet.swift:
```diff
@@ -0,0 +1,581 @@
+  /// - Complexity: O(*n*), where *n* is the number of ranges in the range
+  ///   set.
+  @usableFromInline
+  internal func _inverted<C>(within collection: C) -> RangeSet
```
Candidate for `inlinable` if this is called heavily in other inlinable code.

milseman commented on stdlib/public/core/RangeSet.swift:
```diff
@@ -0,0 +1,581 @@
+internal struct _Pair<Element>: RandomAccessCollection {
+  var pair: (first: Element, second: Element)
+  
+  init(_ first: Element, _ second: Element) {
```
explicit `internal` on all these entities

milseman commented on stdlib/public/core/RangeSetStorage.swift:
```diff
@@ -0,0 +1,169 @@
+//
+//===----------------------------------------------------------------------===//
+
+struct _RangeSetStorage<T: Comparable> {
```
explicit `internal`

milseman reviewed (commented):
Talked with Nate on the phone, recording some points here.

natecook1000 committed:
Revise inlining annotations after discussion w/ @milseman

In particular, only adding inlining when functions can reduce or
eliminate generic parameters. I'll follow up with benchmarks to test
the effects of adding more inlinability.

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 committed:
Add missing internal annotation

natecook1000 committed:
Merge branch 'master' into nc_rangeset

natecook1000 committed:
Update test expectations

With availability annotations, the RangeSet subscript overloads aren't coming
through for these tests.

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 committed:
Add availability to test fixtures

natecook1000 commented:
@swift-ci Please smoke test

drodriguez commented:
@compnerd: I am far from a real computer, but it seems that the Windows failures might simply be fixed moving the `RangeSet.swift` from `test` to `validation-test`. I only see `import StdlibCollectionUnitests` in the `validation-test` directory. I cannot find where that path is set, though.

compnerd commented:
Validation test just effectively disables it (I need to rebase my patch to enable that on the Azure agents).  Could just be something we are setting for that test suite.  I'm waiting for an update to complete; Ill look at it once the update is complete.

compnerd commented:
I think that there is a missing dependency

lugeke commented on stdlib/public/core/RangeSetStorage.swift:
```diff
@@ -0,0 +1,176 @@
+    switch (lhs._storage, rhs._storage) {
+    case (.empty, .empty):
+      return true
+    case (.empty, .singleRange), (.singleRange, .empty):
```
When case .singleRange is empty, doesn’t it equal to case .empty?
